quizzes/quiz6.md

```diff
@@ -0,0 +1,313 @@
+```
+
+The second approach is the new one we are going to discuss today. It is motivated by a concept called **encapsulation**. Basically, a well encapsulated object, does not allow the outside world to change its properties directly, instead it provides a mechanism, so that outsiders can send messages to the object to get certain changes happen. We call this well-defined interaction with the outside world the interface of the object. If we return back to John, we might see code written like this, to achieve the same outcome of marrying Sally:
+
+```js
+john.marryTo("Sally");
+```
+
+If we look at the structure of this code, it has parentheses at the end, which looks like a function call. You are right in your observation. In this indirect method of changing facts about an object, we basically send a message to an object, in the form of a function call. For john to accept this message, it needs to have an implementation. We can implement this function like this:
+
+```js
+let john = {
+    name: "John",
+    isMarried: false,
+    livesIn: "2 bed room flat",
+    marryTo: function(spName) {
+        john.isMarried = true;
+        john.supposesName = spName;
+        john.livesIn = "house";
+        john.numberOfChildren = 0;
+    }
+}
+```
+
+Now our object `john` can happily marry Sally. This implementation has a small flaw which we are going to fix soon, when we talk about **`this`** variable. By the way, a function which is a part of an object property is called a **method**. Worth remembering.
+
+To summarize, we implement functions inside the object definition, to create an encapsulation against the outside world. To call a function which belongs to an object, we need to know the name of the property this function is associated with and also we need to have a variable which references the object we are trying to change.
+
+In JavaScript code, you will see both the direct and indirect way of changing properties of objects. Take the following snippets from our code examples:
+
+```js
+document.body.style.backgroundColor = "blue";
+```
+
+Here we are directly modifying the backgroundColor property of the body's style property.
+
+```js
+document.body.appendChild(productElement);
+```
+
+Here, we are using an indirect method of adding a new element to our dom.
+
+---
+
+**this variable**
+
+We said that we have a flaw in our code:
+
+```js
+let john = {
+    name: "John",
+    isMarried: false,
+    livesIn: "2 bed room flat",
+    marryTo: function(spName) {
+        john.isMarried = true;
+        john.supposesName = spName;
+        john.livesIn = "house";
+        john.numberOfChildren = 0;
+    }
+}
+```
+
+Imagine we have Adam, who wants to marry to Jane. To do that, we do the following:
+
+```js
+let adam = john;
+john = null;
+adam.name = "Adam";
+adam.marryTo("Jane");
+```
+
+We will get an error on the last line. Since marryTo is trying to use an object which does not exist anymore, it is null. So basically it is trying to run this:
```

Comment:
```suggestion
We will get an error on the last line. Since marryTo is trying to use an object which does not exist anymore, it is undefined. So basically it is trying to run this:
```